Update linuxgeneric.rb

falkheiland opened the pull request:
edited regex for prompt to work on SLES 12-SP4

## Pre-Request Checklist
<!-- Not all items apply to each PR, but a great PR addresses all applicable items. -->

- [ ] Passes rubocop code analysis (try `rubocop --auto-correct`)
- [ ] Tests added or adapted (try `rake test`)
- [ ] Changes are reflected in the documentation
- [ ] User-visible changes appended to [CHANGELOG.md](/CHANGELOG.md)

## Description
wanted to use system:

```
NAME="SLES"
VERSION="12-SP4"
VERSION_ID="12.4"
PRETTY_NAME="SUSE Linux Enterprise Server 12 SP4"
ID="sles"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:12:sp4"
```

with `linuxgeneric.rb`

had to edit line2 
`prompt /^(\w.*|\W.*)(:|#) /`
 to:  
`prompt /^(\w.*|\W.*)(:|#) */`
to work

wk commented:
Hi! Could you provide an example of the prompt that necessitates this change? The default shell in SLES 12 should be `bash`, and the shell should normally look like `hostname#`, so the requirement to allow arbitrary characters after the bash prompt is unexpected. 

Are you accessing this host over SSH or Telnet?

falkheiland commented:
Will check it next week.

falkheiland commented:
- using ssh to connect.
- other devices with CentOS7 and 8, Debian , Ubuntu 1804 work using `/opt/rh/rh-ruby23/root/usr/local/share/gems/gems/oxidized-0.27.0/lib/oxidized/model/linuxgeneric.rb` ootb
-  2x SLES 12-SP4 server (came as appliances, so the could not be ootb regarding prompt) do not work

## Prompt

- bash:
SUSE:~ # cat /etc/os-release

## prompt /^(\w.*|\W.*)(:|#) */

- working

```
{:name=>"SUSE",
 :full_name=>"SUSE",
 :ip=>"192.168.1.1",
 :group=>nil,
 :model=>"LinuxGeneric",
 :last=>
  {:start=>2020-08-31 10:19:47 UTC,
   :end=>2020-08-31 10:19:53 UTC,
   :status=>:success,
   :time=>6.543236564},
 :vars=>nil,
 :mtime=>2020-08-31 04:09:26 UTC}
```

```
D, [2020-08-31T12:29:43.076175 #20533] DEBUG -- : lib/oxidized/worker.rb: Added /SUSE to the job queue
D, [2020-08-31T12:29:43.076213 #20533] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:29:43.076275 #20533] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for SUSE at 2020-08-31 10:29:43 UTC
D, [2020-08-31T12:29:43.076429 #20533] DEBUG -- : lib/oxidized/input/ssh.rb: Connecting to SUSE
D, [2020-08-31T12:29:43.076547 #20533] DEBUG -- : AUTH METHODS::["none", "publickey", "password"]
D, [2020-08-31T12:29:43.183959 #20533] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^(\w.*|\W.*)(:|#) */] at SUSE
D, [2020-08-31T12:29:43.987315 #20533] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login commands at SUSE
D, [2020-08-31T12:29:43.987439 #20533] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login command: nil, block: #<Proc:0x00007fb3041743f0@/opt/rh/rh-ruby23/root/usr/local/share/gems/gems/oxidized-0.27.0/lib/oxidized/model/linuxgeneric.rb:69> at SUSE
D, [2020-08-31T12:29:43.987570 #20533] DEBUG -- : lib/oxidized/model/model.rb Collecting commands' outputs
D, [2020-08-31T12:29:43.987610 #20533] DEBUG -- : lib/oxidized/model/model.rb Executing cat /etc/hostname
D, [2020-08-31T12:29:43.987645 #20533] DEBUG -- : lib/oxidized/input/ssh.rb cat /etc/hostname @ SUSE with expect: /^(\w.*|\W.*)(:|#) */
```


## prompt /^(\w.*|\W.*)(:|#) / 

- NOT working

```
 {:name=>"SUSE",
 :full_name=>"SUSE",
 :ip=>"192.168.1.1",
 :group=>nil,
 :model=>"LinuxGeneric",
 :last=>nil,
 :vars=>nil,
 :mtime=>"unknown"}
```

```
D, [2020-08-31T12:31:43.614382 #20591] DEBUG -- : lib/oxidized/input/ssh.rb: Connecting to SUSE
D, [2020-08-31T12:31:43.614547 #20591] DEBUG -- : AUTH METHODS::["none", "publickey", "password"]
D, [2020-08-31T12:31:43.722196 #20591] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^(\w.*|\W.*)(:|#) /] at SUSE
D, [2020-08-31T12:31:44.525611 #20591] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login commands at SUSE
D, [2020-08-31T12:31:44.525929 #20591] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login command: nil, block: #<Proc:0x00007feccc15ba10@/opt/rh/rh-ruby23/root/usr/local/share/gems/gems/oxidized-0.27.0/lib/oxidized/model/linuxgeneric.rb:69> at SUSE
D, [2020-08-31T12:31:44.526015 #20591] DEBUG -- : lib/oxidized/model/model.rb Collecting commands' outputs
D, [2020-08-31T12:31:44.526082 #20591] DEBUG -- : lib/oxidized/model/model.rb Executing cat /etc/hostname
D, [2020-08-31T12:31:44.526115 #20591] DEBUG -- : lib/oxidized/input/ssh.rb cat /etc/hostname @ SUSE with expect: /^(\w.*|\W.*)(:|#) /
D, [2020-08-31T12:31:44.526226 #20591] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^(\w.*|\W.*)(:|#) /] at SUSE
D, [2020-08-31T12:31:44.614069 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:45.614451 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:46.614961 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:47.615490 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:48.615959 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:49.616485 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:50.617050 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:51.617578 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:52.618026 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:53.618463 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:54.618907 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:55.619330 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:56.619778 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:57.620271 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:58.620714 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:31:59.621172 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:32:00.621590 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:32:01.621945 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:32:02.622379 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:32:03.622818 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
W, [2020-08-31T12:32:04.527098 #20591]  WARN -- : 192.168.1.1 raised Timeout::Error with msg "execution expired"
D, [2020-08-31T12:32:04.527250 #20591] DEBUG -- : lib/oxidized/node.rb: Oxidized::SSH failed for SUSE
D, [2020-08-31T12:32:04.527304 #20591] DEBUG -- : lib/oxidized/job.rb: Config fetched for SUSE at 2020-08-31 10:32:04 UTC
W, [2020-08-31T12:32:04.623553 #20591]  WARN -- : /SUSE status no_connection, retry attempt 1
D, [2020-08-31T12:32:04.623731 #20591] DEBUG -- : lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 0 of 1
D, [2020-08-31T12:32:04.623933 #20591] DEBUG -- : lib/oxidized/worker.rb: Added /SUSE to the job queue
D, [2020-08-31T12:32:04.624192 #20591] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for SUSE at 2020-08-31 10:32:04 UTC
D, [2020-08-31T12:32:04.624274 #20591] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-08-31T12:32:04.624499 #20591] DEBUG -- : lib/oxidized/input/ssh.rb: Connecting to SUSE
D, [2020-08-31T12:32:04.624920 #20591] DEBUG -- : AUTH METHODS::["none", "publickey", "password"]
D, [2020-08-31T12:32:04.736738 #20591] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^(\w.*|\W.*)(:|#) /] at SUSE
D, [2020-08-31T12:32:05.540490 #20591] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login commands at SUSE
D, [2020-08-31T12:32:05.540703 #20591] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login command: nil, block: #<Proc:0x00007feccc0391a0@/opt/rh/rh-ruby23/root/usr/local/share/gems/gems/oxidized-0.27.0/lib/oxidized/model/linuxgeneric.rb:69> at SUSE
D, [2020-08-31T12:32:05.540842 #20591] DEBUG -- : lib/oxidized/model/model.rb Collecting commands' outputs
D, [2020-08-31T12:32:05.540889 #20591] DEBUG -- : lib/oxidized/model/model.rb Executing cat /etc/hostname
D, [2020-08-31T12:32:05.540916 #20591] DEBUG -- : lib/oxidized/input/ssh.rb cat /etc/hostname @ SUSE with expect: /^(\w.*|\W.*)(:|#) /
D, [2020-08-31T12:32:05.541109 #20591] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^(\w.*|\W.*)(:|#) /] at SUSE
```

davama commented:
Working?
http://regex.observepoint.com/?regex=%5E(%5Cw.*%7C%5CW.*)(%3A%7C%23)%20&test=SUSE%3A~%20%23%20cat%20%2Fetc%2Fos-release

falkheiland commented:
yes, this works (also with copied / pasted prompt from the sles machine in question).

davama commented:
But for some reason it does not work on your machine...

I would check the PS1 prompt.
Rewrite it and see. It could be that you have some terminal bell character or something.

falkheiland commented:
output of the SLES machine of PS1

```
SUSE:~ # echo $PS1
\[\]\h:\w #\[\]
```

I then created a ~/bashrc with the PS1 from a working CentOS machine:

```
PS1="[\u@\h:\l \W]\\$ "
```

now the original PROMPT `prompt /^(\w.*|\W.*)(:|#) /` works.

I am not sure if the above mentioned SLES prompt is a standard one, which in turn might make it worth to take action in the linuxgeneric.rb.

mortzu commented:
Please process the checklist
